File: appserver/modules/sitehopping_without_single_node/sitehopping_without_single_node.py
import cherrypy
import controllers.module as module

import splunk, splunk.search, splunk.util, splunk.entity
import json
from splunk.appserver.mrsparkle.lib import jsonresponse
import lib.util as util
import lib.i18n as i18n

import logging
logger = logging.getLogger('splunk.module.sitehopping_without_single_node')

import math
import cgi
import os
import re

def node_already_exists(container, c):
    for i in container:
        if i['name'] == str(c):
            return True
    return False


def id_of_node(container, c):
    for i in container:
        if i['name'] == c:
            return i['id']
    return False



class sitehopping_without_single_node(module.ModuleHandler):

    def generateResults(self, host_app, client_app, sid, count=1000, 
            offset=0, entity_name='results'):

        count = max(int(count), 0)
        offset = max(int(offset), 0)
        if not sid:
            raise Exception('hpviz-sitehopping_without_single_node.generateResults - sid not passed!')

        try:
            job = splunk.search.getJob(sid)
        except splunk.ResourceNotFound, e:
            logger.error('hpviz-sitehopping_without_single_node could not find job %s. Exception: %s' % (sid, e))
            return _('<p class="resultStatusMessage">Could not get search data.</p>')
        dataset = getattr(job, entity_name)[offset: offset+count]
        fieldNames = [x for x in getattr(job, entity_name).fieldOrder if (not x.startswith('_'))]
        events = []
        for i, result in enumerate(dataset):
            event = {}
            for field in fieldNames:
                event[field] = str(result.get(field, None))
                if event[field] == 'None':
                    event.pop(field)
                events.append(event)
        lines = []
        for event in events:
            line = event['object2']
            lines.append(line)
        links = []
        i = 0
        nodes = []
        for l in lines:
            lsplit = l.split('-->')
            if lsplit[0] == '':
                continue
            link = {}
            if len(lsplit) == 2:
                if lsplit[0] == 'None':
                    continue
                    """
                    if not node_already_exists(nodes, lsplit[1]):
                        node = {}
                        node['name'] = lsplit[1]
                        node['groupName'] = 'Landding site'
                        node['group'] = 1
                        node['id'] = i
                        node['size'] = 300
                        i = i + 1
                        nodes.append(node)
                    """
                else:
                    if not node_already_exists(nodes, lsplit[0]):
                        node = {}
                        node['name'] = lsplit[0]
                        node['groupName'] = 'Landding site'
                        node['group'] = 1
                        node['id'] = i
                        node['size'] = 300
                        i = i + 1
                        nodes.append(node)
                    if not node_already_exists(nodes, lsplit[1]):
                        node = {}
                        node['name'] = lsplit[1]
                        node['groupName'] = 'Hopping site'
                        node['group'] = 2
                        node['id'] = i
                        node['size'] = 400
                        i = i + 1
                        nodes.append(node)
                    link['source'] = id_of_node(nodes, lsplit[0])
                    link['target'] = id_of_node(nodes, lsplit[1])
                    link['value'] = 1
                    links.append(link)
            elif len(lsplit) == 1:
                if not node_already_exists(nodes, lsplit[0]):
                    node = {}
                    node['name'] = lsplit[0]
                    node['groupName'] = 'Single link'
                    node['group'] = 3
                    node['id'] = i
                    node['size'] = 500
                    i = i + 1
                    nodes.append(node)
    
        for n in nodes:
            if not re.search(r'[:/"]', n['name']):
                if re.search(r'[a-z0-9]{32}', n['name']):
                    n['group'] = 4
                    n['groupName'] = 'Malware md5'
                    n['size'] = 500
        outputJSON = {}

        nodes.sort(key=lambda x:x['id'])
        outputJSON['nodes'] = nodes
        outputJSON['links'] = links
        cherrypy.response.headers['Content-Type'] = 'text/json'
        return json.dumps(outputJSON, sort_keys=True)
        #return self.render_json(outputJSON)
        
    def render_json(self, response_data, set_mime='text/json'):
        cherrypy.response.headers['Content-Type'] = set_mime

        if isinstance(response_data, jsonresponse.JsonResponse):
            response = response_data.toJson().replace("</", "<\\/")
        else:
            response = json.dumps(response_data).replace("</", "<\\/")

        # Pad with 256 bytes of whitespace for IE security issue. See SPL-34355
        return ' ' * 256  + '\n' + response

File: appserver/modules/sitehopping_without_single_node/sitehopping_without_single_node.js
Splunk.Module.sitehopping_without_single_node = $.klass(Splunk.Module.DispatchingModule, {
    initialize: function($super, container) {
        $super(container);
	this.myParam = this.getParam("myParam");
        this.resultsContainer = this.container;
    },
	
    onJobDone: function(event) {
        this.getResults();
    },

    getResultParams: function($super) {
        var params = $super();
        var context = this.getContext();
        var search = context.get("search");
        var sid = search.job.getSearchId();

        if (!sid) this.logger.error(this.moduleType, "Assertion Failed.");

        params.sid = sid;
        return params;
    },

    renderResults: function($super, results) {
        if(!results) {
            this.resultsContainer.html("No content available.");
            return;
        }
        console.debug('Get data');
        if (!d3.select("div.sitehopping_without_single_node_malware > svg").empty()) {
            d3.select("div.sitehopping_without_single_node_malware svg").remove();
        }
        var node,
			link,
			root = results,
			width = 600,
			height = 800;

		root.fixed = true;
		root.x = 0;
		root.y = 0;

	    var force = d3.layout.force()
			.on("tick", tick)
			.charge(function(d) {return -10;})
			.linkDistance(function(d) {return 20})
			.size([width, height]);

		var vis = d3.select("div.sitehopping_without_single_node_malware").append("svg")
			.attr("width", width)
			.attr("height", height);
		update();
		function update() {
			var nodes = root.nodes,
				links = root.links;
		force
			.nodes(nodes)
			.links(links)
		    .start();

		var path = vis.append("svg:g").selectAll("path")
		    .data(force.links())
		  .enter().append("svg:path")
		    .attr("class", function(d) { return "link " + d.type; })
		    .attr("marker-end", function(d) { return "url(#" + d.type + ")"; });


  		link = vis.selectAll("line.link")
	    	.data(links, function(d) { return d.target.id; });

		link.enter().insert("line", ".node")
			.attr("class", "link")
			.attr("x1", function(d) { return d.source.x; })
			.attr("y1", function(d) { return d.source.y; })
			.attr("x2", function(d) { return d.target.x; })
			.attr("y2", function(d) { return d.target.y; });

		// Exit any old links.
		link.exit().remove();

		// Update the nodes…
		node = vis.selectAll("circle.node")
		    .data(nodes, function(d) { return d.id; })
		    .style("fill", color);


		node.transition()
		    .attr("r", function(d) { return d.children ? 7.5 : d.size / 100; });

		// Enter any new nodes.
		node.enter().append("circle")
		    .attr("class", "node")
		    .attr("cx", function(d) { return d.x; })
		    .attr("cy", function(d) { return d.y; })
		    .attr("r", function(d) { return d.children ? 7.5 : d.size / 100; })
		    .style("fill", color)
		    .on("click", click)
		    .call(force.drag);
		node.append("title")
			.text(function(d) {
          var d = this.__data__;
          var description;
          if (d.group == 1) {
              description = 'Landding site : ';
          }
          else if (d.group == 2) {
              description = 'Hopping site : ';
          }
          else if (d.group == 3) {
              description = 'Single Link : ';
          }
          else if (d.group == 4) {
              description = 'malware md5 : ';
          }
          else
              description = 'Undefined : ';
          var append = d.country ? ' location : ' + d.country : '';

        return description + d.name + append;
    });


  // Exit any old nodes.
  node.exit().remove();
}
function tick() {
  link.attr("x1", function(d) { return d.source.x; })
      .attr("y1", function(d) { return d.source.y; })
      .attr("x2", function(d) { return d.target.x; })
      .attr("y2", function(d) { return d.target.y; });

  node.attr("cx", function(d) { return d.x; })
      .attr("cy", function(d) { return d.y; });
}
function color(d) {
    if (d.group == 0) {
        return "#CC33CC";
    }
    else if (d.group == 1) {
        return "#00EECC";
    }
    else if (d.group == 2) {
        return "#000000";
    }
    else if (d.group == 3) {
        return "#0000EE";
    }
    else if (d.group == 4) {
        return "#FF0000";
    }
    else
        return "#EE0000";
//  return d._group ? "#00ee00" : d.children ? "#000000" : "#0000ee";
}
function click(d) {
}
/*

        var w = 1400,
            h = 900;
        var root = results;
        root.fix = true;
        root.x = w / 2;
        root.y = h / 2;
        update();

        function update() {
            var nodes = root.nodes;
            var links = root.links;


            var force = d3.layout.force()
                .nodes(nodes)
                .links(links)
                .size([w, h])
                .linkDistance(45)
                .charge(-25)
                .on("tick", tick)
                .start();

            var svg = d3.select("div.sitehoppint_malware").append("svg:svg")
                .attr("width", w)
                .attr("height", h);

            var path = svg.append("svg:g").selectAll("path")
                .data(force.links())
              .enter().append("svg:path")
                .attr("class", function(d) { return "link " + d.type; })
                .attr("marker-end", function(d) { return "url(#" + d.type + ")"; });

            var circle = svg.append("svg:g").selectAll("circle")
                .data(force.nodes())
              .enter().append("svg:circle")
                .attr("r", 6)
                .call(force.drag)
                .style("fill", color);
            circle.append("title")
                .text(function(d) {return d.name});
                
            function tick() {
                path.attr("d", function(d) {
                        var dx = d.target.x - d.source.x,
                        dy = d.target.y - d.source.y,
                        dr = Math.sqrt(dx * dx + dy * dy);
                        return "M" + d.source.x + "," + d.source.y + "A" + dr + "," + dr + " 0 0,1 " + d.target.x + "," + d.target.y;
                        });

                circle.attr("transform", function(d) {
                        return "translate(" + d.x + "," + d.y + ")";
                        });
            }
            function color(d) {
                if (d.group == 1)
                    return "red";
                if (d.group == 2)
                    return "blue";
                if (d.group == 3)
                    return "silver";
                if (d.group == 4)
                    return "black";
                else
                    return "yellow";
            }
        }
*/
	}
});
